File: src/main/java/com/github/frcsty/frozenactions/wrapper/ActionHandler.java
package com.github.frcsty.frozenactions.wrapper;

import com.github.frcsty.frozenactions.actions.Action;
import com.github.frcsty.frozenactions.actions.ActionContext;
import com.github.frcsty.frozenactions.actions.player.AdventureSoundAction;
import com.github.frcsty.frozenactions.actions.broadcast.*;
import com.github.frcsty.frozenactions.actions.player.*;
import com.github.frcsty.frozenactions.time.TimeAPI;
import me.mattstudios.msg.adventure.AdventureMessage;
import me.mattstudios.msg.base.MessageOptions;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.bukkit.plugin.Plugin;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.*;
import java.util.concurrent.TimeUnit;
import java.util.logging.Level;
import java.util.logging.Logger;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public final class ActionHandler {

    private static final Pattern ACTION_PATTERN = Pattern.compile("(.*) ?\\[(?<action>[A-Z]+?)] ?(?<arguments>.+)", Pattern.CASE_INSENSITIVE);
    private static final Pattern DELAY_PATTERN = Pattern.compile("\\[DELAY=(?<delay>\\d+[a-z])]", Pattern.CASE_INSENSITIVE);
    private static final Pattern CHANCE_PATTERN = Pattern.compile("\\[CHANCE=(?<chance>\\d+)]", Pattern.CASE_INSENSITIVE);
    private static final SplittableRandom RANDOM = new SplittableRandom();

    private static AdventureMessage adventureMessage = AdventureMessage.create();

    private final Map<String, Action> actions = new HashMap<>();
    private final Plugin plugin;
    private final Logger logger;

    public ActionHandler(final Plugin plugin) {
        this.plugin = plugin;
        this.logger = plugin.getLogger();

        Bukkit.getMessenger().registerOutgoingPluginChannel(plugin, "BungeeCord");
    }

    public static AdventureMessage getAdventureMessage() {
        return adventureMessage;
    }

    public void loadDefaults(final boolean defaultAdventureMessage) {
        Arrays.asList(
                // Broadcast
                new AdventureSoundBroadcastAction(),
                new MiniMessageBroadcastAction(),
                new ActionbarBroadcastAction(),
                new BroadcastAction(),
                new CenterBroadcastAction(),
                new JsonBroadcastAction(),
                new SoundBroadcastAction(),
                new TitleBroadcastAction(),

                // Player
                new AdventureSoundAction(),
                new MiniMessageAction(),
                new ActionbarMessageAction(),
                new BungeeAction(),
                new CenterMessageAction(),
                new ConsoleCommandAction(),
                new JsonMessageAction(),
                new MessageAction(),
                new PlayerCommandAction(),
                new ChatAction(),
                new SoundAction(),
                new TeleportAction(),
                new TitleMessageAction(),
                new GameModeAction(),
                new PotionEffectAction(),
                new TimeAction(),
                new ParticleAction(),
                new HealthAction(),
                new FoodLevelAction()
        ).forEach(it -> actions.put(it.getId().toUpperCase(), it));

        if (defaultAdventureMessage)
            adventureMessage = AdventureMessage.create();
    }

    public void createAdventureMessage(final @NotNull MessageOptions options) {
        adventureMessage = AdventureMessage.create(options);
    }

    public void setAction(final @NotNull Action action) {
        this.actions.put(action.getId().toUpperCase(), action);
    }

    public void removeAction(final @NotNull String identifier) {
        this.actions.remove(identifier.toUpperCase());
    }

    public void execute(final @NotNull Player player, final @NotNull List<String> actions) {
        actions.forEach(it -> execute(player, it));
    }

    private void execute(final @NotNull Player player, final @NotNull String input) {
        if (hasChanceAction(input) == null) return;
        final String chanceAction = hasChanceAction(input);

        if (chanceAction == null) return;
        final ActionHolder actionHolder = getDelayAction(chanceAction);
        final String inputAction = actionHolder.getAction();
        final Matcher match = ACTION_PATTERN.matcher(inputAction);

        if (!match.matches()) {
            logger.info("Action does not match regex '" + inputAction + "'!");
            return;
        }

        final String arguments = match.group("arguments");
        final String actionName = match.group("action").toUpperCase();
        final long delay = actionHolder.getDelay();

        final Action action = this.actions.get(actionName);

        if (action == null) return;

        Bukkit.getScheduler().runTaskLater(
                plugin,
                () -> action.run(new ActionContext(plugin, player, arguments)),
                delay
        );
    }

    private @Nullable String hasChanceAction(@NotNull String input) {
        final Matcher match = CHANCE_PATTERN.matcher(input);

        if (!match.matches())
            return input;

        final String chanceGroup = match.group("chance");

        if (chanceGroup == null)
            return null;

        final int chance = Integer.parseInt(chanceGroup);
        final int randomValue = RANDOM.nextInt(100) + 1;

        return randomValue <= chance ? input.replace(chanceGroup, "") : null;
    }

    private @NotNull ActionHolder getDelayAction(final @NotNull String input) {
        final Matcher match = DELAY_PATTERN.matcher(input);

        if (!match.matches())
            return new ActionHolder(input, 0L);

        final String delayGroup = match.group("delay");

        if (delayGroup == null)
            return new ActionHolder(input, 0L);

        try {
            final TimeAPI.TimeResult time = new TimeAPI().parseTime(delayGroup);

            return new ActionHolder(
                    input.replace(delayGroup, ""),
                    time.to(TimeUnit.SECONDS) * 20
            );
        } catch (final IllegalArgumentException ex) {
            logger.log(Level.SEVERE, "", ex);
        }

        return new ActionHolder(input, 0L);
    }

}

File: src/main/kotlin/com/github/frcsty/frozenactions/wrapper/ActionHandler.kt
package com.github.frcsty.frozenactions.wrapper

import com.github.frcsty.frozenactions.actions.Action
import com.github.frcsty.frozenactions.actions.broadcast.*
import com.github.frcsty.frozenactions.actions.player.*
import com.github.frcsty.frozenactions.time.parseTime
import org.bukkit.Bukkit
import org.bukkit.entity.Player
import org.bukkit.plugin.Plugin
import java.util.*
import java.util.concurrent.TimeUnit

private val ACTION_PATTERN = Regex("(.*) ?\\[(?<action>[A-Z]+?)] ?(?<arguments>.+)", RegexOption.IGNORE_CASE)
private val DELAY_PATTERN = Regex("\\[DELAY=(?<delay>\\d+[a-z])]", RegexOption.IGNORE_CASE)
private val CHANCE_PATTERN = Regex("\\[CHANCE=(?<chance>\\d+)]", RegexOption.IGNORE_CASE)
private val RANDOM = SplittableRandom()

class ActionHandler(private val plugin: Plugin) {

    val actions = mutableMapOf<String, Action>()

    fun loadDefault() {
        setOf(
                ActionbarBroadcastAction,
                ActionbarMessageAction,
                BroadcastAction,
                BungeeAction,
                CenterBroadcastAction,
                CenterMessageAction,
                ConsoleCommandAction,
                EquipItemAction,
                JsonBroadcastAction,
                JsonMessageAction,
                MessageAction,
                PlayerCommandAction,
                SoundAction,
                SoundBroadcastAction,
                TeleportAction,
                TitleBroadcastAction,
                TitleMessageAction
        ).forEach { this.actions[it.id] = it }
    }

    fun setAction(identifier: String, action: Action) {
        this.actions[identifier] = action
    }

    fun removeAction(identifier: String) {
        this.actions.remove(identifier)
    }

    fun execute(player: Player, input: List<String>) {
        input.forEach { execute(player, it) }
    }

    private fun execute(player: Player, input: String) {
        val actionHolder = getDelayAction(hasChanceAction(input) ?: return)
        val inputAction = actionHolder.action
        val match = ACTION_PATTERN.matchEntire(inputAction)

        if (match == null) {
            println("Action does not match regex $inputAction")
            return
        }

        val arguments = match.groups["arguments"]?.value ?: return
        val delay = actionHolder.delay

        val actionName = match.groups["action"]?.value?.toUpperCase()
        val action = actions[actionName] ?: return

        Bukkit.getScheduler().runTaskLater(
                plugin,
                Runnable {
                    action.run(player, arguments)
                    action.run(plugin, player, arguments)
                },
                delay
        )
    }

    private fun hasChanceAction(input: String): String? {
        val match = CHANCE_PATTERN.matchEntire(input) ?: return input
        val chanceGroup = match.groups["chance"] ?: return null
        val chance = chanceGroup.value.toInt()
        val randomValue = RANDOM.nextInt(100) + 1

        return if (randomValue <= chance) input.replace(chanceGroup.value, "") else null
    }

    private fun getDelayAction(input: String): ActionHolder {
        val match = DELAY_PATTERN.matchEntire(input) ?: return ActionHolder(input, 0L)

        val delayGroup = match.groups["delay"] ?: return ActionHolder(input, 0L)
        val delay = delayGroup.value

        return try {
            val time = delay.parseTime()
            ActionHolder(
                    action = input.replace(delayGroup.value, ""),
                    delay = time.to(TimeUnit.SECONDS) * 20L
            )
        } catch (ex: IllegalArgumentException) {
            ex.printStackTrace()
            ActionHolder(input, 0L)
        }
    }

    init {
        Bukkit.getMessenger().registerOutgoingPluginChannel(plugin, "BungeeCord")
    }
}